app/src/main/java/emu/skyline/provider/SkylineDocumentsProvider.kt

```diff
@@ -0,0 +1,307 @@
+        result = includeFile(result, documentId, null)
+        return result
+    }
+
+    override fun isChildDocument(parentDocumentId : String?, documentId : String?) : Boolean {
+        return documentId?.startsWith(parentDocumentId!!) ?: false
+    }
+
+    override fun createDocument(parentDocumentId : String?, mimeType : String?, displayName : String) : String? {
+        var newFile = File(parentDocumentId, displayName)
+        var noConflictId = 2
+        while (newFile.exists()) {
+            newFile = File(parentDocumentId, displayName + " (" + noConflictId++ + ")")
+        }
+        try {
+            val succeeded : Boolean = if (DocumentsContract.Document.MIME_TYPE_DIR == mimeType) {
+                newFile.mkdir()
+            } else {
+                newFile.createNewFile()
+            }
+            if (!succeeded) {
+                throw FileNotFoundException("Failed to create document with id " + newFile.path)
+            }
+        } catch (e : IOException) {
+            throw FileNotFoundException("Failed to create document with id " + newFile.path)
+        }
+        return newFile.path
+    }
+
+    override fun deleteDocument(documentId : String?) {
+        val file = getFileForDocId(documentId!!)
+        if (!file.delete()) {
+            throw FileNotFoundException("Couldn't delete document with documentId $documentId")
+        }
+    }
+
+    override fun removeDocument(documentId : String, parentDocumentId : String?) {
+
+        val parent = getFileForDocId(parentDocumentId!!)
+        val file = getFileForDocId(documentId) ?: throw FileNotFoundException("Failed to delete document with id $documentId")
+
+
+        var doesFileParentMatch = false
+        val fileParent = file.parentFile
+        if (fileParent == null || fileParent.equals(parent)) {
+            doesFileParentMatch = true
+        }
+
+        if (parent.equals(file) || doesFileParentMatch) {
+            if (file.delete()) {
+
+            } else {
+                throw FileNotFoundException("Failed to delete document with id $documentId")
+            }
+        } else {
+            throw FileNotFoundException("Failed to delete document with id $documentId")
+        }
+    }
+
+    override fun renameDocument(documentId : String?, displayName : String?) : String? {
+
+        if (displayName == null) {
+            throw FileNotFoundException("Couldn't rename document, new name is null")
+        }
+
+        val sourceFile = getFileForDocId(documentId!!)
+        val sourceParentFile = sourceFile.parentFile ?: throw FileNotFoundException("Failed to rename document. File has no parent.")
+        val destFile = File(sourceParentFile.path, displayName)
+
+        try {
+            val renameSucceeded = sourceFile.renameTo(destFile)
+            if (!renameSucceeded) {
+                throw FileNotFoundException("Failed to rename document. Renamed failed.")
+            }
```

Comment:
Just move the call inside the `if` statement, I will say that it is annoying that this exception is thrown and caught by the outer handler even though I get why.